### PR TITLE
Handwritten parser for parsing requirements

### PR DESCRIPTION
**Pull request by hrnciar**
Hello,

I wrote a parser for packaging that should allow us to remove the dependency on pyparsing. 
I am opening this as a draft as this is my first PR into packaging and I would love to hear feedback from you on what to improve in my PR. 

Thank you.

Fixes: #468

### CODE REVIEW

**Comment by brettcannon** [comment]
Two quick things (I haven't really reviewed any of the code).

One, looks like there are some CI failures.

Two, do we want `packaging.parser` and `packaging.tokenizer` to be public? They strike me as implementation details, so we would probably prefer to keep them as private (i.e. leading underscore in the module names).

**Comment by pradyunsg** [comment]
+1 to both the things that Brett said.

**Comment by pradyunsg** [comment]
All that said, this is super cool and I'm very excited to see this progressing! Thanks for picking this up @hrnciar! ^>^

File: packaging/requirements.py
Comment (pradyunsg) [reply packaging/requirements.py:37]: Yea. It would also make sense to make this a private module-level global, so... `_RequirementTuple` works too. :)

File: packaging/_parser.py
Comment (uranusjr) [inline packaging/_parser.py:11]: What does the `strip("'\"")` part do? I don't think the NAME part can contain quotes. Should this be `strip()`?

Comment (hrnciar) [reply packaging/_parser.py:11]: You are right, the `strip()` is unnecessary and can be removed entirely.

File: packaging/_parser.py
Comment (hroncok) [inline packaging/_parser.py:6]: Shouldn't this actually be:

```suggestion
def parse_named_requirement(requirement: str) -> Tuple[str, str, str, str, str]:
```

? Similarily in all the other places where the type is set to Any.

Comment (hrnciar) [reply packaging/_parser.py:6]: In this case, it should be `Tuple[str, str, List[str], str, str]`. I will take a look at the rest of them.

File: packaging/_parser.py
Comment (pradyunsg) [inline packaging/_parser.py:1]: I'd appreciate it if we can get rid of as many of the `Any`s here as feasible. Not a blocker, obviously.

File: packaging/_parser.py
Comment (pradyunsg) [inline packaging/_parser.py:117]: We should move the AST classes to over here, allowing them to be used here without needing a deferred import to avoid import cycles.

File: packaging/_tokenizer.py
Comment (pradyunsg) [inline packaging/_tokenizer.py:74]: Isn't this guaranteed to be a token?

File: packaging/_tokenizer.py
Comment (pradyunsg) [inline packaging/_tokenizer.py:77]: It's likely reasonable to require source to be a string.

File: packaging/_tokenizer.py
Comment (pradyunsg) [inline packaging/_tokenizer.py:85]: These none annotations are weird. If it's not supposed to be passed, just don't have varargs/kwargs?

File: packaging/_tokenizer.py
Comment (pradyunsg) [inline packaging/_tokenizer.py:126]: The message should likely not have a default value, and instead be a required kwarg.

File: packaging/_tokenizer.py
Comment (pradyunsg) [reply packaging/_tokenizer.py:85]: This basically applies to the whole file.

File: packaging/_tokenizer.py
Comment (pradyunsg) [inline packaging/_tokenizer.py:8]: This is missing a -> None

File: packaging/_parser.py
Comment (pradyunsg) [inline packaging/_parser.py:6]: It would be nice to return a namedtuple from here, so that this definition is self-contained.

**Comment by pradyunsg** [comment]
Pulling this out of draft status, given there is a review request.

**Review by pradyunsg — COMMENTED** [review]
This is pretty much a desk review -- I haven't cloned and poked at this yet; I'll do that once the linters are happy with this PR.

I don't think I have anything to say about the parsing logic other than that it looks correct, and that CI being happy is a good sign. :)

I think we can/should bundle this change with the legacy version removal, which... I'll pick back up once this lands.

File: packaging/_parser.py
Comment (hroncok) [reply packaging/_parser.py:1]: I know that @hrnciar is working on redoing the typing annotations not to use Anys. Most of the typing annotation comments from this review will be resolved by his WIP changes.

Comment (pradyunsg) [reply packaging/_parser.py:1]: And I imagine our mypy lint job is complaining about them as well. :)

**Comment by pradyunsg** [comment]
A gentle ping.

File: packaging/_parser.py
Comment (encukou) [inline packaging/_parser.py:36]: These names are a bit confusing. Perhaps use `MarkerAtom`, `MarkerItem`, `MarkerVar` and `MarkerList`, corresponding with their parse rules?

**Comment by pradyunsg** [comment]
(FYI: The CI isn't running due to merge conflicts with `main`)

File: packaging/_parser.py
Comment (hrnciar) [reply packaging/_parser.py]: Yes, you are correct. I replaced it with `str`. Thank you.

**Comment by hrnciar** [comment]
> (FYI: The CI isn't running due to merge conflicts with `main`)

Thanks, I rebased PR on top of the main branch.

File: packaging/_parser.py
Comment (encukou) [inline packaging/_parser.py:86]: This looks like it'll will allow the comma to be skipped. Please test `spam[eggs ham]`: it should not be valid.

File: packaging/_parser.py
Comment (encukou) [inline packaging/_parser.py:106]: Similar situation here.

File: packaging/_parser.py
Comment (encukou) [inline packaging/_parser.py:114]: This function should have a docstring with the grammar.

File: packaging/_parser.py
Comment (encukou) [inline packaging/_parser.py:157]:
```suggestion
    MARKER_VAR: VARIABLE | MARKER_VALUE
```

File: packaging/_parser.py
Comment (encukou) [inline packaging/_parser.py:180]:
```suggestion
        python_str = tokens.read().text[1:-1]
```

File: packaging/_tokenizer.py
Comment (encukou) [inline packaging/_tokenizer.py:13]: The `text` argument is now unused, can you remove it?

 If yes, `match_args`/`match_kwargs` below can be replaced by `name`.

File: packaging/_tokenizer.py
Comment (encukou) [inline packaging/_tokenizer.py:34]:
```suggestion
    "COMMA": r",",
```

File: packaging/_tokenizer.py
Comment (encukou) [inline packaging/_tokenizer.py:106]: It turns out that the autogenerated  error messages aren't really used. It might be better if `expect` and `read` took an `error_message` argument.

File: tests/test_requirements.py
Comment (encukou) [inline tests/test_requirements.py:106]: Can the `value` attribute be added back? (Is it really missing?)

File: tests/test_requirements.py
Comment (encukou) [inline tests/test_requirements.py:70]: These specifications should still be tested. Do they cause the right error?

File: tests/test_requirements.py
Comment (encukou) [inline tests/test_requirements.py:93]: The old specifications should still be tested. Do they cause the right error?

File: packaging/_parser.py
Comment (uranusjr) [reply packaging/_parser.py:180]: Would it make sense to use `ast.literal_eval` to handle escape sequences? (I think they are technically allowed...? PEP 345 simply says the value can be “a free string”.)

Comment (encukou) [reply packaging/_parser.py:180]: [PEP 508](https://www.python.org/dev/peps/pep-0508/#grammar) specifies a *very* restricted charset for this, actually. But the old parser uses pyparsing's QuotedString, so `ast.literal_eval` would match the old behavior better.
(That is what we want -- where possible, any behavior changes to better match the spec should not be part of *this* change.)

File: tests/test_requirements.py
Comment (hrnciar) [reply tests/test_requirements.py:93]: I've added a `test_missing_name` to cover this. Yes, it causes the right error.

File: packaging/requirements.py
Comment (pradyunsg) [inline packaging/requirements.py:37]: This should be computed outside of `__init__` to avoid paying the cost of creating this class each time an instance is created.

File: packaging/_parser.py
Comment (uranusjr) [inline packaging/_parser.py:12]:
```suggestion
        return self.value
```

Since `self.value` is always str

File: packaging/_parser.py
Comment (pradyunsg) [inline packaging/_parser.py:147]: not: drop the comma.

File: packaging/_parser.py
Comment (pradyunsg) [inline packaging/_parser.py:169]: nit: drop new line at start of function.

**Review by pradyunsg — APPROVED** [review]
LGTM, with only nit picks now.

There’s a few follow ups, like improving the error message on a semicolon that’s missing white space and on legacy versions; but we should do those in a follow up.

A blocker for releasing this is going to be packaging dropping the legacy versions and legacy specifiers though, so… that needs to be done as well. I’ll pick up #407 sometime in the coming week or next year.

File: packaging/_parser.py
Comment (pradyunsg) [reply packaging/_parser.py:169]: Same for the other instances of this.

**Comment by hrnciar** [comment]
I've pushed a change to address the nit picks. Thank you all for your help with this PR.

**Comment by ssbarnea** [comment]
@hrnciar Any chance to revive/rebase this? I think that this patch is becoming more important as py311 is getting closer. Thanks!

**Comment by hroncok** [comment]
We were told this is blocked on https://github.com/pypa/packaging/pull/407

File: packaging/specifiers.py
Comment (pradyunsg) [inline packaging/specifiers.py:350]: AHAHAHAH. We've both spent way-too-long on this.

File: packaging/_parser.py
Comment (pradyunsg) [inline packaging/_parser.py:91]:
```suggestion
            if tokens.match("RBRACKET"):
```

We don't want to allow `,,,` in extras? This should get a corresponding test too, IMO.

**Review by pradyunsg — CHANGES_REQUESTED** [review]
A few more suggestions. 😅

File: packaging/_parser.py
Comment (pradyunsg) [reply packaging/_parser.py:91]: Ah, nvm. I misread this. :)

File: packaging/_tokenizer.py
Comment (pradyunsg) [inline packaging/_tokenizer.py:131]: I really like this! :)

File: packaging/specifiers.py
Comment (hrnciar) [reply packaging/specifiers.py:404]: I split the regex into two parts so I could reuse it in the tokenizer for the VERSION token.
~~~~
"VERSION": re.compile(Specifier._version_regex_str, re.VERBOSE | re.IGNORECASE),
~~~~

**Review by brettcannon — CHANGES_REQUESTED** [review]
One of the few things that Black can't check syntactically, the docstrings unfortunately don't match the style in the rest of the project (both the newline structure and the sentences need to end with proper punctuation).

**Comment by hrnciar** [comment]
I've addressed some remarks, but there is still some stuff to do. I am heading to EuroPython tomorrow so I will continue next week when I am back. If anyone would like to meet in Dublin and have a chat IRL feel free to ping me :).

**Review by pradyunsg — APPROVED** [review]
LGTM, assuming @encukou's suggestion is applied!

**Comment by hrnciar** [comment]
I've improved the docstrings and fixed a couple of nitpicks. Unless there is something else to improve, I think, we are done.

**Comment by hrnciar** [comment]
@brettcannon could you please do a review? Is there something specific I should improve? I'd like to finish this PR. Thank you.

**Comment by brettcannon** [comment]
@hrnciar it's in my review queue, but if @pradyunsg wants to merge w/o a review for me then I am totally fine with that.

**Comment by hroncok** [comment]
Changelog entry: https://github.com/pypa/packaging/pull/581

**Comment by ofek** [comment]
This is great for CLIs too 🙂 https://github.com/pypa/hatch/issues/449#issuecomment-1236125448

**Pull request merged**
